otel-allocator/allocation/allocator.go

```diff
@@ -0,0 +1,130 @@
+package allocation
+
+import (
+	"log"
+	"sync"
+)
+
+/*
+	Load balancer will serve on an HTTP server exposing /jobs/<job_id>/targets <- these are configured using least connection
+	Load balancer will need information about the collectors in order to set the URLs
+	Keep a Map of what each collector currently holds and update it based on new scrape target updates
+*/
+
+// Create a struct that holds collector - and jobs for that collector
+// This struct will be parsed into endpoint with collector and jobs info
+
+type collector struct {
+	Name       string
+	NumTargets int
+}
+
+// allocator makes decisions to distribute work among
+// a number of OpenTelemetry collectors based on the number of targets.
+// Users need to call SetTargets when they have new targets in their
+// clusters and call Reshard to process the new targets and reshard.
+
+type Allocator struct {
+	m     sync.Mutex
+	cache displayCache
+
+	targetsWaiting map[string]TargetItem // temp buffer to keep targets that are waiting to be processed
+
+	collectors map[string]*collector // all current collectors
+
+	nextCollector *collector
+	targetItems   map[string]*TargetItem
+}
+
+// findNextCollector finds the next collector with less number of targets.
+func (allocator *Allocator) findNextCollector() {
+	for _, v := range allocator.collectors {
+		if v.NumTargets < allocator.nextCollector.NumTargets {
+			allocator.nextCollector = v
+		}
+	}
+}
+
+// SetTargets accepts the a list of targets that will be used to make
+// load balancing decisions. This method should be called when where are
+// new targets discovered or existing targets are shutdown.
+func (allocator *Allocator) SetTargets(targets []TargetItem) {
+	// Dump old data
+	allocator.m.Lock()
+	for k := range allocator.targetsWaiting {
+		delete(allocator.targetsWaiting, k)
+	}
+	// Set new data
+	for _, i := range targets {
+		allocator.targetsWaiting[i.JobName+i.TargetURL] = i
+	}
+	allocator.m.Unlock()
+}
+
+// SetCollectors sets the set of collectors with key=collectorName, value=Collector object.
+func (allocator *Allocator) SetCollectors(collectors []string) {
+	// TODO: How do we handle the new collectors?
+
+	if len(collectors) == 0 {
+		log.Fatal("no collector instances present")
+	}
+	allocator.m.Lock()
+	for _, i := range collectors {
```

Comment:
Does the collector map also need to be cleared here?  What happens when a collector disappears?  I don't think this would remove it from the list.

Reply:
For now, when there is a change with the collectors, we are doing a full re-balance. SetCollectors() will be called with a brand new list of collectors and these collectors are all guaranteed to be up.